servers/rendering/renderer_rd/shaders/effects/screen_space_reflection.glsl

```diff
@@ -219,16 +220,40 @@ void main() {
 			//margin_blend = 1.0;
 		}
 
-		vec2 final_pos;
+		// Fade In / Fade Out
 		float grad = (steps_taken + 1.0) / float(params.num_steps);
 		float initial_fade = params.curve_fade_in == 0.0 ? 1.0 : pow(clamp(grad, 0.0, 1.0), params.curve_fade_in);
 		float fade = pow(clamp(1.0 - grad, 0.0, 1.0), params.distance_fade) * initial_fade;
+
+		// Ensure that precision errors do not introduce any fade. Even if it is just slightly below 1.0,
+		// strong specular light can leak through the reflection.
+		if (fade > 0.999) {
+			fade = 1.0;
+		}
```

Comment:
This is the only change in behaviour in this PR right? Everything else is just a cosmetic change?

Reply:
This and we also need to multiply Fresnel with final_color.rgb instead of final_color.a.
So in the original version change
`final_color.a *= f0 + (1.0 - f0) * m;`
to
`final_color.rgb *= f0 + (1.0 - f0) * m;`
Sorry for mixing cosmetic changes with changes in behaviour.

Reply:
Hmm, that seems incorrect to me. 

``final_color.rgb`` represents the diffuse color while ``final_color.a`` represents the specular blend factor. The fresnel factor is a part of calculating specular reflections, but not diffuse lighting. This change makes the fresnel factor change diffuse lighting, but not specular. 

That is why the specular is still leaking so bad when fade is not 0. 

I think the real issue is that the specular highlight is so bright that it isn't blending nicely. In other words, the specular highlight is so bright that no matter what the blend factor is, it shows up. This can happen when diffuse is in a normal range (0-10 or so) and specular is in a much higher range (100s). Commonly this occurs when using a raw HDRI that wasn't designed for game engines

Reply:
Inspecting the HDRI in gimp confirms my suspicions. The center of the sun goes to a brightness of about 32,000 units. Any amount of blending will result in that bleeding through. E.g. mix(sun, SSR, 0.01) Still results in the sun contribution being 320 units.

Reply:
I agree that the extremely bright sun is the main issue here. However, I still think that Fresnel in this specific case needs to be combined with `final_color.rgb` and that Godot could at least handle such an HDRi well without fade in/out.

This is main() in specular_merge.glsl (not affected by this PR; `final_color` is `ssr_color` here):
https://github.com/godotengine/godot/blob/eca6f0eb545166708c56633152974f40f51de2be/servers/rendering/renderer_rd/shaders/effects/specular_merge.glsl#L94-L107
I believe there are three types of light here:
- General specular light (from the sky, lights),
- SSR specular light (final_color/ssr_color),
- General diffuse light (unrelated to SSR output; from the sky, lights).

I agree that `final_color.a` represents the blend factor (of general specular and SSR specular).
However, from my understanding, `final_color.rgb` represents the **specular reflection** (of the diffuse light coming from the reflected object). We treat the incoming (previously diffuse) light as specular which is why we need to apply Fresnel to it.

If the reflections were diffuse, then every point on the mirror would look the same from every angle and we know that is not the case with SSR. I believe that diffuse reflections are simulated by SDFGI.

Behavior currently in master without this change:
1. Darken (apply Fresnel to) general specular light in the PBR shader.
2. Blend SSR specular with general specular based on Fresnel and fades in specular_merge. 

Behavior in this PR:
1. Darken (apply Fresnel to) general specular light in the PBR shader.
2. Darken (apply Fresnel to) SSR specular light in the SSR shader.
3. Blend SSR specular with general specular based on fades in specular_merge.

Also, I cannot see an obvious way to blend low range with high range specular as we cannot use tonemapping at this stage. But the fade in/out could be refactored so it stays at 1.0 instead of at a problematic value like 0.999 and only starts dropping exponentially when it reaches a distance threshold.

Reply:
Hmmm, yes, thank you for your detailed explanation. You are right, since the main scene specular has Fresnel already applied, it is reasonable that the SSR specular should also have fresnel pre-applied. Using fresnel as a blend factor means that we are introducing light in places where the reflection is simply meant to be weaker.